pkg/segment/reader/segread/timereader.go

```
@@ -434,6 +440,11 @@ func ReadAllTimestampsForBlock(blkNums map[uint16]struct{}, segKey string,
 			allReadJob <- &timeBlockRequest{tsRec: rawBlock, blkNum: currBlk, numRecs: numRecs}
 			readOffset += readLen
 		}
+		err = segreader.PutBufToPool(rawChunk)
```

`rawBlock` (line 440) is a slice of `rawChunk`. You need to wait until the message you put on the channel on line 440 is done with `rawBlock` before you can put `rawChunk` back in the pool, to avoid memory corruption

Before your changes, `rawChunk` was put back in a pool under a `defer` statement, which was safe because it happens after the `readerWG.Wait()` call on line 450

---

Can you suggest a good pattern for handling that without ignoring the error and 'defer'ing it still?

---

A few options that come to mind:
1. Once you're done with this `rawChunk` send a special message on the channel. That message indicates to the worker receiving it that it should `PutBufToPool(rawChunk)`—so then you do the PutBuf in a worker and not here.
2. You could keep track of how many messages you put on the channel for each `rawChunk`, then have each worker increment an atomic number when they're done processing messages. Then once that atomic counter reaches a threshold, you know it's safe to `PutBufToPool(rawChunk)`. Not sure what a good way to monitor the counter would be though (maybe callbacks?).

Of those options, I think option 1 will be a bit more clean, but they were just the first things that came to mind; maybe you'll have a better idea.

I'm assuming that once a message on the channel is fully processed, that part of `rawChunk` is never used again. I think that's right, but you should double check.

I'm just realizing you might be asking "how can I defer it but still handle the error". If that's what you're asking, see https://go.dev/play/p/4T3Zpw85Zj2. You could `err = PutBufToPool(rawChunk); log(err)` inside the closure

---

It was indeed the later. Fixed.